### PR TITLE
[google_maps_flutter] Add background color for unloaded tiles

### LINKED ISSUE #142520: [google_maps_flutter] Background color for unloaded tiles
### Use case

When using a dark style for the google map, a white background is being shown for the unloaded tiles.

### Proposal

It would be a good idea to have at disposal a variable like backgroundColor defined in the MapConfiguration class which can be used to change the color for the unloaded tiles depending on the app theme.

I've seen that on android, there is a variable like this defined in the GoogleMapOptions class:
https://developers.google.com/android/reference/com/google/android/gms/maps/GoogleMapOptions#backgroundColor(java.lang.Integer)

Also for ios:
https://developers.google.com/maps/documentation/ios-sdk/reference/interface_g_m_s_map_view_options#a929091a8ea8a162385cf7f0dfe0a666f

### PR DESCRIPTION
**Pull request by RyanHolanda**
Adds a nullable `GoogleMap.backgroundColor` option for configuring the color displayed behind map tiles while they are not fully loaded.

This prevents the native map surface from showing its default black background before tiles render. The value is forwarded during map initialization to:

- Android's `GoogleMapOptions.backgroundColor`.
- iOS's `GMSMapViewOptions.backgroundColor` in the default, SDK 9, and SDK 10 implementations.
- Web's JavaScript `MapOptions.backgroundColor`.

The existing SDK defaults remain unchanged when `backgroundColor` is null. Since the native options are initialization-time settings, changing the value after map creation has no effect.

No screenshots are included because this is an API-only platform integration change.

Fixes https://github.com/flutter/flutter/issues/142520

## Federated review note

This is the required combined-review PR for the [federated plugin change]. It intentionally contains repository-generated path dependency overrides marked `DO NOT MERGE` so the platform interface, implementations, and app-facing package can be reviewed and tested together.

The package publish dry-run passes for all seven packages. The federation safety check intentionally reports `Unresolved combo PR`, preventing this combined review branch from landing accidentally. Once the overall change is approved, it will be split and landed in dependency order:

1. Platform interface.
2. Android, iOS, and web implementations.
3. App-facing package after the preceding releases are published.

## Test coverage

The change adds coverage for:

- App-facing configuration forwarding.
- Platform-interface diffing, equality, hashing, and legacy serialization.
- Android Dart-to-Pigeon conversion and native creation options.
- iOS Dart-to-Pigeon conversion and native `GMSMapViewOptions` creation across all maintained SDK implementations.
- Web conversion to the JavaScript map creation options, including alpha.

Local validation completed:

- Repository formatter: passed.
- Dart analysis and tests for all seven packages: passed.
- Android Gradle lint and native unit tests: passed.
- iOS SDK 9 and SDK 10 analysis and native tests: passed (72 tests each).
- iOS shared-source synchronization validator: passed.
- Web integration tests against the latest and minimum supported Flutter versions: passed.
- Package validation, license validation, and publish dry-runs: passed.

On the local Xcode 26.5 installation, the default CocoaPods native target is blocked before plugin tests by warnings promoted to errors in the third-party `Google-Maps-iOS-Utils` source. The identical shared implementation passes in the SDK 9 and SDK 10 targets; hosted CI will validate the repository's supported CocoaPods environment.

## Pre-Review Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [AI contribution guidelines] and understand my responsibilities, or I am not using AI tools.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran [the auto-formatter].
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I followed [the version and CHANGELOG instructions], using [semantic versioning] and the [repository CHANGELOG style], or I have commented below to indicate which documented exception this PR falls under[^1].
- [x] I updated/added any relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or I have commented below to indicate which [test exemption] this PR falls under[^1].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

[^1]: Regular contributors who have demonstrated familiarity with the repository guidelines only need to comment if the PR is not auto-exempted by repo tooling.

[federated plugin change]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changing-federated-plugins
[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[AI contribution guidelines]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#ai-contribution-guidelines
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[the auto-formatter]: https://github.com/flutter/packages/blob/main/script/tool/README.md#format-code
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[the version and CHANGELOG instructions]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version-and-changelog-updates
[semantic versioning]: https://dart.dev/tools/pub/versioning#semantic-versions
[repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[test exemption]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart [hunk 1]
```diff
@@ -1267,6 +1267,9 @@ PlatformMapConfiguration _platformMapConfigurationFromMapConfiguration(MapConfig
     markerType: _platformMarkerTypeFromMarkerType(config.markerType ?? MarkerType.marker),
     mapId: config.mapId,
     style: config.style,
+    backgroundColor: config.backgroundColor == null
+        ? null
+        : PlatformColor(argbValue: config.backgroundColor!.toARGB32()),
```
Comment (stuartmorgan-g) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart]: Prefer creating a local `backgroundColor` variable and using that, so that you can rely on compiler-enforced autopromotion instead of force-unwrapping.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart [hunk 2]
```diff
@@ -1310,6 +1313,9 @@ PlatformMapConfiguration _platformMapConfigurationFromOptionsJson(Map<String, Ob
     markerType: PlatformMarkerType.marker,
     mapId: options['mapId'] as String?,
     style: options['style'] as String?,
+    backgroundColor: options['backgroundColor'] == null
+        ? null
+        : PlatformColor(argbValue: options['backgroundColor']! as int),
```
Comment (stuartmorgan-g) [inline packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart]: Same.

File: packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_maps_flutter_ios.dart [hunk 3]
```diff
@@ -1162,6 +1162,9 @@ PlatformMapConfiguration _platformMapConfigurationFromMapConfiguration(MapConfig
     markerType: _platformMarkerTypeFromMarkerType(config.markerType ?? MarkerType.marker),
     mapId: config.mapId,
     style: config.style,
+    backgroundColor: config.backgroundColor == null
+        ? null
+        : _platformColorFromColor(config.backgroundColor!),
```
Comment (stuartmorgan-g) [inline packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_maps_flutter_ios.dart]: Same notes here as for Android.

**Review by stuartmorgan-g — COMMENTED** [review]
Overall the structure looks good, thanks for the contribution!

Routing to platform teams for in-depth review of platform implementations before the PR is split apart.

**Review by mdebbar — COMMENTED** [review]
The web changes are straightforward and look good to me!

**Review by mdebbar — APPROVED** [review]

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart (hunk 1 above)
Comment (RyanHolanda) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart]: Fixed in d318693b56. I now store `config.backgroundColor` in a local `Color?` and use the promoted local for the conversion, removing the force unwrap.

File: packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart (hunk 2 above)
Comment (RyanHolanda) [reply packages/google_maps_flutter/google_maps_flutter_android/lib/src/google_maps_flutter_android.dart]: Fixed in d318693b56. The legacy options path now stores `options["backgroundColor"]` in a local `int?` and uses compiler promotion instead of force-unwrapping the map lookup.

File: packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_maps_flutter_ios.dart (hunk 3 above)
Comment (RyanHolanda) [reply packages/google_maps_flutter/google_maps_flutter_ios/lib/src/google_maps_flutter_ios.dart]: Fixed in d318693b56. I applied the same promoted-local pattern to both analogous iOS conversion paths. Android/iOS analysis and all package and example Dart tests pass.

**Review by vashworth — APPROVED** [review]
iOS side LGTM

**Comment by vashworth** [comment]
@RyanHolanda Can you resolve the merge conflicts so we can get the tests to run?

**Comment by RyanHolanda** [comment]
@vashworth Resolved, thanks! I rebased the PR onto the current `main` and kept both the upstream `2.16.1` release entry and this PR’s `2.17.0` entry. The PR is mergeable now and fresh CI checks are queued.

Locally, all seven affected packages pass `analyze`, `dart-test`, `validate`, and `publish-check`.

**Comment by RyanHolanda** [comment]
Resolved again. The new conflicts came from #12307 converting the maintained SDK9/SDK10/shared iOS controller code to Swift; that upstream change did not duplicate this PR’s `backgroundColor` feature.

I rebased onto current `main` and ported only this PR’s iOS initialization to the new Swift `GoogleMapController`: `GMSMapViewOptions.backgroundColor` is set from the creation configuration before constructing `GMSMapView`. I also added the equivalent Swift test, updated the generated-code paths for the new `_objc` targets, and left the frozen legacy iOS implementation in Objective-C.

Validation passes for all affected packages (`analyze`, `dart-test`, `validate`, and `publish-check`), shared SDK source synchronization passes, and the SDK10 `GoogleMapsTests` suite ran on a physical iPhone on iOS 26.6: 9/9 passed, including `mapViewOptionsUseConfiguredBackgroundColor()`.

@vashworth Could you please take another look?

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/GoogleMapBuilderTest.java [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+
+import com.google.android.gms.maps.GoogleMapOptions;
+import org.junit.Test;
+
+public class GoogleMapBuilderTest {
+
+  @Test
+  public void setBackgroundColorConfiguresMapCreationOptions() {
```
Comment (reidbaker) [inline packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/GoogleMapBuilderTest.java:16]: did you want to add a test for the null behavior since that was the main feature you were trying to add?

**Review by reidbaker — COMMENTED** [review]
android code looks good % a null test.

File: packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/GoogleMapBuilderTest.java (hunk 4 above)
Comment (RyanHolanda) [reply packages/google_maps_flutter/google_maps_flutter_android/android/src/test/java/io/flutter/plugins/googlemaps/GoogleMapBuilderTest.java:16]: Added in 0a46410cbe. Since the null check is in GoogleMapFactory (before the value reaches GoogleMapBuilder), I added a factory-level test that passes a configuration with a null background color and verifies that setBackgroundColor is never called, preserving the native default. The full Android native unit-test suite passes.

**Comment by hellohuanlin** [comment]
@RyanHolanda could you resolve the code conflict?

**Comment by RyanHolanda** [comment]
@hellohuanlin Resolved the conflicts against current main and pushed the updated branch. The resolution includes the Swift Pigeon migration for the iOS SDK9/SDK10 implementations and refreshed package versions/changelogs. Analysis, Dart tests, repository validation, publish checks, Android unit tests, and the targeted iOS SDK10 simulator test passed. GitHub now reports the PR as mergeable; CI is pending.
